### ISSUE TITLE
Buffer overflow not properly patched

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#728]
**Reported by opensource on 1 Apr 2010 21:38 UTC**

changeset:1676 seems not to address the recently discovered buffer overflow.

From the Fedora bug report: (Bug 577654 in Red Hat's Bugzilla)

 * The code checks if the self-proclaimed size of the packet is larger than the real packet size. If the packet is larger than 256 bytes AND correctly tells about that, the heap will still be overwritten...
 * The self-proclaimed size of the packet is compared to uninitialized data, resulting in random results
 * They forgot to patch airbase-ng.c

**Comment by aircrack-ng** [comment]
**Comment by misterx on 1 Apr 2010 21:38 UTC**

Fixed now in aircrack-ng/aircrack-ng@cafc49c3234cd95f4f5b4ed9830ca6d9a6874dd3 for aircrack-ng, airdecap-ng and airodump-ng and aircrack-ng/aircrack-ng@9f16b517336a1dc95e425bb4a4779feef6f2e881 for airbase-ng

**Comment by aircrack-ng** [comment]
**Modified by misterx on 1 Apr 2010 21:38 UTC**

**Comment by aircrack-ng** [comment]
**Comment by opensource on 1 Apr 2010 21:38 UTC**

Here is a comment from the original bug reporter claiming that the fix is not yet ok:
"""
I've only checked airodump-ng and as far as I can see the fix is incorrect as
the field "pkh.len" is uninitialized. It just happens to contain a value that
prevents the bug from getting triggered - this is left to a random value on the
stack however. It should read "caplen" instead of "pkh.len".

Why don't you just compare the promoted size of the EAPOL-frame to
"sizeof(wpa.eapol_frame)" ?
"""

**Comment by aircrack-ng** [comment]
**Comment by clopez on 1 Apr 2010 21:38 UTC**

Here are the relevant links:

 * Bug report: https://bugzilla.redhat.com/show_bug.cgi?id=577654
 * Blog post about the issue: https://pyrit.wordpress.com/2010/03/28/remote-exploit-against-aircrack-ng/

I have tested it with trunk (aircrack-ng/aircrack-ng@200263037cfaa1cd04b4449a862778b30a9adeec):

```
svn co http://trac.aircrack-ng.org/svn/trunk/ aircrack-ng
cd aircrack-ng
make
wget http://pyrit.googlecode.com/svn/tags/opt/aircrackng_exploit.cap
```

```
$ aircrack-ng aircrackng_exploit.cap 
Opening aircrackng_exploit.cap
Read 1 packets.

   #  BSSID              ESSID                     Encryption

   1  00:DE:AD:C0:DE:00                            WPA (0 handshake)

Choosing first network as target.

Opening aircrackng_exploit.cap
Reading packets, please wait...
```

It keeps there forever and you have to hit CTR+C twice to make it stop   :\

**Comment by aircrack-ng** [comment]
**Comment by vanhoefm on 1 Apr 2010 21:38 UTC**

Several tools are still vulnerable because it is not checked whether the .eapol_size is small enough. My patch fixes the vulnerability in several tools. Aircrack-ng ran forever because it did not release the mutex before reading the next packet. The WPA_hdsk structure was copied all over the codebase, and is now placed in eapol.h which is included by all tools.

I have a working exploit allowing arbitrary code execution when aircrack-ng loads a specially crafted .pcap file... Remote exploitation of other tools should also be possible. So please accept this patch!

**Comment by aircrack-ng** [comment]
**Comment by misterx on 1 Apr 2010 21:38 UTC**

Fixed in aircrack-ng/aircrack-ng@3d140fed2c76dcb1b5c5c51ccb8753129fe3ac5e and aircrack-ng/aircrack-ng@d7e2399ea949e0e1227c23556710445d4c99fc14.
